File: rfc/rfc/structs/feature.py
from typing import Any
from collections.abc import Iterable
from enum import Enum

from ..types import *

class FeatureType(Enum):
    NUMERICAL = 0
    CATEGORICAL = 1
    BINARY = 2

class Feature(IdentifiedObject):
    name: str
    type: FeatureType
    _levels: None | list[float] = None
    _categories: None | list[Any] = None

    def __init__(
        self,
        id_: int,
        type_: FeatureType,
        name: None | str = None,
        levels: None | Iterable[float] = None,
        categories: None | list[Any] = None,
    ) -> None:
        self.id = id_
        self.name = name if name is not None else f'Feature {id_}'
        self.type = type_
        if levels is not None:
            self.levels = levels
        if categories is not None:
            self.categories = categories

    def isnumerical(self) -> bool:
        return self.type == FeatureType.NUMERICAL

    def isbinary(self) -> bool:
        return self.type == FeatureType.BINARY

    def iscategorical(self) -> bool:
        return self.type == FeatureType.CATEGORICAL

    def value(self, x: Sample) -> Any:
        return x[self.id]

    @property
    def levels(self) -> list[float]:
        if not self.isnumerical():
            raise ValueError("Levels are only defined for numerical features!")
        if self._levels is None:
            raise ValueError("Feature has no levels!")
        return self._levels

    @levels.setter
    def levels(self, levels: Iterable[float]):
        if not self.isnumerical():
            raise ValueError("Levels are only defined for numerical features!")
        self._levels = list(sorted(levels))
                
    @property
    def categories(self) -> list[Any]:
        if not self.iscategorical():
            raise ValueError("Categories are only defined for categorical features!")
        if self._categories is None:
            raise ValueError("Feature has no categories!")
        return self._categories

    @categories.setter
    def categories(self, categories: list[Any]):
        if not self.iscategorical():
            raise ValueError("Categories are only defined for categorical features!")
        self._categories = categories

    def __str__(self) -> str:
        return self.name

    def __repr__(self) -> str:
        return f'{self.__class__.__name__}({self.id}, {self.type})'

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Feature):
            return NotImplemented
        return self.id == other.id

    def __hash__(self) -> int:
        return hash(self.id)

File: rfc/rfc_cpp/struct/src/feature.cpp
#include <iostream>
#include <vector>
#include <variant>
#include "../include/feature.hpp"

Feature::Feature(int id_, FeatureType type_, std::string name = "") {
    this->id = id_;
    this->name = name.empty() ? "Feature " + std::to_string(id_) : name;
    this->type = type_;
}

bool Feature::isnumerical() {
    return this->type == FeatureType::NUMERICAL;
}

bool Feature::isbinary() {
    return this->type == FeatureType::BINARY;
}

bool Feature::iscategorical() {
    return this->type == FeatureType::CATEGORICAL;
}

double Feature::value(std::vector<double> x){
    return x[this->id];
}
/*
void Feature::setLevels(std::vector<float> levels_){
    if (!(this->isnumerical())){
        throw std::invalid_argument("Levels are only defined for numerical features!");
    }
    this->levels = levels_;
}

std::vector<float> Feature::getLevels(){
    if (!(this->isnumerical())){
        throw std::invalid_argument("Levels are only defined for numerical features!");
    }

    return this->levels;
}

std::vector<int> Feature::getCategories(){
    if (!(this->iscategorical())){
        throw std::invalid_argument("Categories are only defined for categorical features!");
    }
    return this->categories;
}

void Feature::setCategories(std::vector<int> categories_){
    if (!(this->iscategorical())){
        throw std::invalid_argument("Categories are only defined for categorical features!");
    }
    this->categories = categories_;
}
*/

bool Feature::operator==(const Feature& other) const {
    return this->id == other.id;
}



namespace std {
    template<>
    struct hash<Feature> {
        size_t operator()(const Feature& obj) const {
            return std::hash<int>()(obj.id); // Use the hash function for int
        }
    };
}
/*
int main() {
    FeatureType type1 = FeatureType::NUMERICAL;
    Feature feature1(1,type1,"ftr1");
    std::cout<<feature1<<std::endl;
    std::cout<<feature1.isnumerical()<<std::endl;
    std::vector<float> nums = {1.f, 2.f, 3.14f, 4.f, 5.f};

    std::cout<<std::endl;
    Feature feature2(2,type1,"ftr2");
    std::cout<<(feature1==feature2)<<std::endl;
    std::cout<<std::hash<Feature>{}(feature1)<<std::endl;

}
*/